### ISSUE TITLE
[JIT] treat `assert not torch.jit.is_scripting()` statically

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#44864]
## 🚀 Feature
JIT should treat `assert not torch.jit.is_scripting()` statically (like how it treats `if is_scripting()` and ignore what's after the assertion in a basicblock.

## Motivation

Having more ways to let torchscript ignore code sections will in general make it easier to migrate code to partially support torchscript.

## Pitch
Currently, this is scriptable:
```
class A(nn.Module):
    def __init__(self):
        super().__init__()
        self.k = 1.0
    def forward(self, x):
        if self.training and not torch.jit.is_scripting():
            return x * self.k if self.k != "test" else x   # not supported due to lack of union type
        else:
            return x + 1

a = A()
a = torch.jit.script(a)
```
but this fails:
```
class A(nn.Module):
    def __init__(self):
        super().__init__()
        self.k = 1.0
    def forward(self, x):
        if self.training:
            assert not torch.jit.is_scripting(), "this codepath is not supported"
            return x * self.k if self.k != "test" else x   # error in torchscript due to type mistmatch
        else:
            return x + 1

a = A()
a = torch.jit.script(a)
```

The former case can be used as a workaround but it's not an ideal solution, because it does not throw errors when user tries to use the scripted model in training mode. Bypassing code sections using `assert not torch.jit.is_scripting()` seems like the simplest syntax: it's only one line of code and does not introduce extra indentation. This is similar to `@torch.jit.unused`: it skips the code with one extra line and throws an error if the skipped code is called.
(This issue is not about union type specifically, since it applies to other python features missing in torchscript.)

## Alternatives
```python
class A(nn.Module):
    def __init__(self):
        super().__init__()
        self.k = 1.0
    def forward(self, x):
        if self.training:
            if not torch.jit.is_scripting():
                return x * self.k if self.k != "test" else x
            else:
                assert False, "this codepath is not supported"
        else:
            return x + 1
```
this works now but it's more ugly.

cc @gmagogsfm
